src/motion/features/viewport/use-viewport.ts

```diff
@@ -0,0 +1,90 @@
+import { MutableRefObject, useEffect, useRef } from "react"
+import { VisualElement } from "../../../render/types"
+import { AnimationType } from "../../../render/utils/types"
+import { FeatureProps } from "../types"
+import { createIntersectionObserver } from "./observers"
+import { ViewportOptions } from "./types"
+
+export function useViewport({
+    visualElement,
+    whileInView,
+    onViewportEnter,
+    onViewportLeave,
+    viewport = {},
+}: FeatureProps) {
+    const hasEnteredView = useRef(false)
+
+    let shouldObserve = Boolean(
+        whileInView || onViewportEnter || onViewportLeave
+    )
+
+    if (viewport.once && hasEnteredView.current) shouldObserve = false
+
+    const useObserver =
+        typeof IntersectionObserver === "undefined"
+            ? useMissingIntersectionObserver
+            : useIntersectionObserver
+
+    useObserver(shouldObserve, hasEnteredView, visualElement, viewport)
+}
+
+function useIntersectionObserver(
+    shouldObserve: boolean,
+    hasEnteredView: MutableRefObject<boolean>,
+    visualElement: VisualElement,
+    { root, margin: rootMargin, amount = "some", once }: ViewportOptions
+) {
+    useEffect(() => {
+        if (!shouldObserve) return
+
+        return createIntersectionObserver(
+            visualElement.getInstance(),
+            {
+                root: root?.current,
+                rootMargin,
+                threshold: amount === "some" ? 0 : 1,
+            },
+            (entry) => {
+                const { isIntersecting } = entry
+
+                if (once && !isIntersecting && hasEnteredView.current) {
```

Comment:
shouldn't this trigger `onViewportLeave`?

Reply:
I'm not sure, but if `once` affects the callbacks then people get that consistent behaviour baked in. It's quite nice for when you wanna trigger other stuff on the first viewport enter